File: ios/EhmSewooBt.m
#import "EhmSewooBt.h"

@implementation EhmSewooBt
RCT_EXPORT_MODULE()

-(void)initialize {
    zplPrinter = [[ZPLPrinter alloc] init];
}

- (NSDictionary *)constantsToExport
{
    
    /*
     EVENT_DEVICE_ALREADY_PAIRED    Emits the devices array already paired
     EVENT_DEVICE_DISCOVER_DONE    Emits when the scan done
     EVENT_DEVICE_FOUND    Emits when device found during scan
     EVENT_CONNECTION_LOST    Emits when device connection lost
     EVENT_UNABLE_CONNECT    Emits when error occurs while trying to connect device
     EVENT_CONNECTED    Emits when device connected
     */

    return @{ @"connecting": @"connecting",
              @"connected":@"connected",
              @"disconnecting":@"disconnecting",
              @"disconnected":@"disconnected",
              @"paperEmpty":@"paperEmpty",
              @"coverOpen":@"coverOpen",
              @"battery":@"battery"
              };
}

- (NSArray<NSString *> *)supportedEvents
{
    /*
     EVENT_DEVICE_ALREADY_PAIRED    Emits the devices array already paired
     EVENT_DEVICE_DISCOVER_DONE    Emits when the scan done
     EVENT_DEVICE_FOUND    Emits when device found during scan
     EVENT_CONNECTION_LOST    Emits when device connection lost
     EVENT_UNABLE_CONNECT    Emits when error occurs while trying to connect device
     EVENT_CONNECTED    Emits when device connected
     */

    return @[@"connecting",
             @"connected",
             @"disconnecting",
             @"disconnected",
             @"paperEmpty",
             @"coverOpen",
             @"battery"];
}

RCT_EXPORT_METHOD(DiscoverDevices:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    resolve(nil);
}

RCT_EXPORT_METHOD(StopDiscover:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    resolve(nil);
}

RCT_EXPORT_METHOD(ConnectDevice:(NSString*)deviceId
                  withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    [self sendEventWithName:@"connecting" body:nil];
    //long ret = [zplPrinter openPort:@"192.168.43.245", withPortParam:9100];
    long ret = [zplPrinter openPort:deviceId withPortParam:9100];
    if(ret >= 0)
    {
        [self sendEventWithName:@"connected" body:nil];
    }
    resolve((@TRUE));
}

RCT_EXPORT_METHOD(GetDevices:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    resolve(nil);
}

RCT_EXPORT_METHOD(Disconnect:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    [self sendEventWithName:@"disconnecting" body:nil];
    [zplPrinter closePort];
    [self sendEventWithName:@"disconnected" body:nil];
    resolve(@(TRUE));
}

RCT_EXPORT_METHOD(PrintZpl:(NSString*)zpl
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    @try {
        if(zplPrinter == nil){
            [self sendEventWithName:@"disconnecting" body:nil];
            resolve(@(FALSE));
        }else {
            
            NSData *data = [zpl dataUsingEncoding:NSUTF8StringEncoding];
            const unsigned char *bytes = (const unsigned char *)[data bytes];
            NSUInteger length = [data length];
            [zplPrinter printData:bytes withLength:length];
            
            resolve(@(TRUE));
        }
    } @catch (NSException *exception)
    {
        [self sendEventWithName:@"disconnecting" body:nil];

        if(!zplPrinter)
        {
            [zplPrinter closePort];
        }
        
        [self sendEventWithName:@"disconnected" body:nil];
        resolve(@(FALSE));
    }
}

- (void) tcpStatusBox:(long) sts
{
    sts = [zplPrinter status];
    if(sts == STS_ZPL_NORMAL)
    {
        
    } else {
       if((sts & STS_ZPL_PAPER_EMPTY) > 0)
       {
           [self sendEventWithName:@"paperEmpty" body:nil];
       }
        if((sts & STS_ZPL_COVER_OPEN) > 0)
        {
            [self sendEventWithName:@"coverOpen" body:nil];
        }
        if((sts & STS_ZPL_BATTERY_LOW) > 0)
        {
            [self sendEventWithName:@"battery" body:nil];
        }
        [self sendEventWithName:@"disconnecting" body:nil];
        [zplPrinter closePort];
        [self sendEventWithName:@"disconnected" body:nil];
    }
}
- (void) statusCheckReceived:(NSNotification *) notification
{
    long bytesAvailable = 0;
    long readLength = 0;
    unsigned char buf[8] = {0,};
    EABluetoothPort * sessionController = (EABluetoothPort *)[notification object];
    NSString * result = [[NSString alloc] init];
#ifdef DEBUG
    NSLog(@"===== Status Check START =====");
#endif
    NSMutableData * readData = [[NSMutableData alloc] init];
    while((bytesAvailable = [sessionController readBytesAvailable]) > 0)
    {
        NSData * data = [sessionController readData:bytesAvailable];
        if(data)
        {
            [readData appendData:data];
            readLength = readLength + bytesAvailable;
        }
    }
    if(readLength > sizeof(buf))
        readLength = sizeof(buf);
    [readData getBytes:buf length:readLength];
    
    int sts = buf[readLength - 1];
    if(sts == STS_ZPL_NORMAL)
    {
       
    }
    else
    {
        if((sts & STS_ZPL_BUSY) > 0)
        {
            result = [result stringByAppendingString:@"Printer Busy\r\n"];
        }
        if((sts & STS_ZPL_COVER_OPEN) > 0)
        {
            result = [result stringByAppendingString:@"Cover Open\r\n"];
        }
        if((sts & STS_ZPL_PAPER_EMPTY) > 0)
        {
            result = [result stringByAppendingString:@"Paper Empty\r\n"];
        }
        if((sts & STS_ZPL_BATTERY_LOW) > 0)
        {
            result = [result stringByAppendingString:@"Battery Low\r\n"];
        }
    }
#ifdef DEBUG
    NSLog(@"===== Status Check EXIT =====");
#endif
}


@end

File: ios/EhmSewooBt.swift
import React
import Foundation
import ExternalAccessory

@objc(EhmSewooBt)
class EhmSewooBt: NSObject, RCTBridgeModule {
    static func moduleName() -> String! {
        return "EhmSewooBt"
    }
    
    var stsData = Data()
    let zplPrinter = ZPLPrinter()
    var media_type:String = ZPL_SENSE_GAP
    static let shared = EhmSewooBt()

    override init() {
        super.init()
        
        let accessoryManager = EAAccessoryManager.shared()
        accessoryManager.registerForLocalNotifications() // bildirimleri almak için kaydol
        
        NotificationCenter.default.addObserver(self, selector: #selector(statusCheckReceived(_:)), name: NSNotification.Name.EADSessionDataReceived, object: nil)
           
    }
    
    @objc func statusCheckReceived(_ notification:Notification)
    {
        let sessionController:EABluetoothPort = notification.object as! EABluetoothPort
        var bytesAvailable:CLong = 0
        
        while(true)
        {
            bytesAvailable = sessionController.readBytesAvailable()
            
            if(bytesAvailable > 0)
            {
                let data:Data = sessionController.readData(UInt(bytesAvailable))
                stsData.append(data)
            }
            else
            {
                break;
            }
        }
        
        if stsData.count > 0 && stsData.isEmpty == false {
            if(stsData[0] == STS_ZPL_NORMAL)
            {
                //alertMessageBox(title: "Printer Status", message: "Normal")
            }
            else {
                var result:String = ""
                var ists:Int = 0
                
                if((stsData[0] & UInt8(STS_ZPL_BUSY)) > 0)
                {
                    result += "Printer Busy\r\n"
                    ists = 1;
                }
                
                if((stsData[0] & UInt8(STS_ZPL_COVER_OPEN)) > 0)
                {
                    result += "Cover Open\r\n";
                    ists = 1;
                }
                
                if((stsData[0] & UInt8(STS_ZPL_PAPER_EMPTY)) > 0)
                {
                    result += "Paper empty\r\n";
                    ists = 1;
                }
                
                if((stsData[0] & UInt8(STS_ZPL_BATTERY_LOW)) > 0)
                {
                    result += "Battery Low\r\n"
                }
                
                
                if( ists == 1)
                {
                    //alertMessageBox(title: "Printer Status", message: result)
                }
            }
            stsData.removeAll()
        }
    }
    
    @objc(DiscoverDevices:withRejecter:)
    func DiscoverDevices(resolve:RCTPromiseResolveBlock, reject: RCTPromiseResolveBlock) -> Void {
        resolve(true)
    }
    
    @objc(StopDiscover:withRejecter:)
    func StopDiscover(resolve:RCTPromiseResolveBlock, reject: RCTPromiseResolveBlock) -> Void {
        resolve(true)
    }
    
    @objc(ConnectDevice:withRejecter:)
    func ConnectDevice(resolve:RCTPromiseResolveBlock, reject: RCTPromiseResolveBlock) -> Void {
        var errCode: Int = 0
        errCode = zplPrinter.openPort("bluetooth", withPortParam: 0)
        
        if(errCode >= 0)
        {
            resolve(true)
        }
        else
        {
            resolve(false)
        }
    }
    
    @objc(GetDevices:withRejecter:)
    func GetDevices(resolve:RCTPromiseResolveBlock, reject: RCTPromiseResolveBlock) -> Void {
        resolve(true)
    }
    
    @objc(Disconnect:withRejecter:)
    func Disconnect(resolve:RCTPromiseResolveBlock, reject: RCTPromiseResolveBlock) -> Void {
        resolve(true)
    }
    
    @objc(PrintZpl:withResolver:withRejecter:)
    func PrintZpl(zpl: String, resolve:RCTPromiseResolveBlock, reject: RCTPromiseResolveBlock) -> Void {
        zplPrinter.setupPrinter(ZPL_ROTATION_180, withmTrack:media_type, withWidth:384, withHeight:480)
        zplPrinter.startPage()
        
        zplPrinter.setInternationalFont(0)
        
        zplPrinter.printText(ZPL_FONT_A, withOrientation:ZPL_ROTATION_0, withWidth:15, withHeight:12, withPrintX:0, withPrintY:0, withData:"FontA 0123")
        zplPrinter.printText(ZPL_FONT_B, withOrientation:ZPL_ROTATION_0, withWidth:15, withHeight:12, withPrintX:0, withPrintY:30, withData:"FontB 0123")
        zplPrinter.printText(ZPL_FONT_C, withOrientation:ZPL_ROTATION_0, withWidth:15, withHeight:12, withPrintX:0, withPrintY:60, withData:"FontC 0123")
        zplPrinter.printText(ZPL_FONT_D, withOrientation:ZPL_ROTATION_0, withWidth:15, withHeight:12, withPrintX:0, withPrintY:90, withData:"FontD 0123")
        zplPrinter.printText(ZPL_FONT_E, withOrientation:ZPL_ROTATION_0, withWidth:15, withHeight:12, withPrintX:0, withPrintY:120, withData:"FontE 0123")
        zplPrinter.printText(ZPL_FONT_F, withOrientation:ZPL_ROTATION_0, withWidth:15, withHeight:12, withPrintX:0, withPrintY:160, withData:"FontF 0123")
        zplPrinter.printText(ZPL_FONT_G, withOrientation:ZPL_ROTATION_0, withWidth:15, withHeight:12, withPrintX:0, withPrintY:210, withData:"FontG 01")
        zplPrinter.printText(ZPL_FONT_H, withOrientation:ZPL_ROTATION_0, withWidth:15, withHeight:12, withPrintX:0, withPrintY:300, withData:"FontH 01234567")

        zplPrinter.endPage(1)
    }
    
}
